### PR TITLE
SVG Charts for PDFs (OX-9272)

### PR DESCRIPTION
**Pull request by jakobvogel**
### Description

Provides a framework for charts that can be rendered (via SVG) into PDFs. Currently, only pie and spider charts are supported.

### Screenies

Screenshot 2024-12-05 at 02 54 49
Screenshot 2024-12-05 at 02 54 36

### Additional Notes

- This PR fixes or works on following ticket(s): [OX-9272](https://scireum.myjetbrains.com/youtrack/issue/OX-9272)

### Checklist

- [x] Code change has been tested and works locally
- [x] Code was formatted via IntelliJ and follows SonarLint & [best practices](https://scireum.myjetbrains.com/youtrack/articles/MISC-A-16/CodeStyle-JavaDoc)

### CODE REVIEW

File: src/main/java/sirius/biz/charts/PieChart.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        }
+
+        return piecePath;
+    }
+
+    private String assemblePathDefinitionForSlice(double startRadians, double endRadians, double radius) {
+
+        // if the start and end point are precisely the same, we need to add a small offset to avoid rendering issues
+        double delta = endRadians - startRadians;
+        double offset = (Math.abs(delta % TWO_PI) < 2.0e-4) ? 1.0e-4 : 0.0;
```
Comment (sabieber) [inline src/main/java/sirius/biz/charts/PieChart.java:171]: **Optional:** Wouldn't `0.0002` and `0.0001` be more readable and with the same number of characters?

File: src/main/java/sirius/biz/charts/SpiderChart.java [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+/**
+ * Represents a spider chart (radar chart) which can be rendered as SVG.
+ *
+ * @param <N> the type of the numeric values
+ */
+public class SpiderChart<N extends Number> extends BaseChart {
```
Comment (sabieber) [inline src/main/java/sirius/biz/charts/SpiderChart.java:33]: **Optional:** Would use the more common name "radar chart" as the class name

File: src/main/java/sirius/biz/charts/SpiderChart.java [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+
+    private static final double TICK_LENGTH = 2.0;
```
Comment (sabieber) [inline src/main/java/sirius/biz/charts/SpiderChart.java:35]: **Optional:** I do miss the units with these constants a bit :)

File: src/main/java/sirius/biz/charts/SpiderChart.java [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+            int datasetIndex = datasetCounter.getAndIncrement();
+            StringBuilder pathDefinition = new StringBuilder();
+
+            dataset.withColor(Dataset.COLORS.get(datasetIndex % Dataset.COLORS.size()));
+
+            AtomicInteger labelCounter = new AtomicInteger();
+            labels.forEach(label -> {
+                int labelIndex = labelCounter.getAndIncrement();
+
+                double radians = 2.0 * Math.PI * labelIndex / labels.size();
```
Comment (sabieber) [inline src/main/java/sirius/biz/charts/SpiderChart.java:195]: **Optional:** Here no constant for tau is used :) Which is also present in the Java standard library as `Math.TAU`

**Review by sabieber — APPROVED** [review]
Looking very nice indeed! 👌🏻
My suggestions are just nitpicking ;)

**Review by idlira — APPROVED** [review]

File: src/main/java/sirius/biz/charts/PieChart.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    private static final double TWO_PI = 2.0 * Math.PI;
+
+    private static final double RING_WIDTH = 4.0;
+
+    private Dataset<N> dataset;
+
+    /**
+     * Flag to determine if the pie chart should be rendered as a ring rather than filled.
+     */
+    private boolean ring = false;
```
[the remark this answers is not in the document]
Comment (jakobvogel) [reply src/main/java/sirius/biz/charts/PieChart.java]: Refactored this into a suitably named sub-class in 6991f3ff3e0e1057225c58b59a19fe10520a521f. Using the spelling from [Wikipedia](https://en.wikipedia.org/wiki/Pie_chart#Doughnut_chart), though.

File: src/main/java/sirius/biz/charts/BaseChart.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+    /**
+     * Contains the tag name for circle elements, {@value}.
+     */
+    protected static final String TAG_CIRCLE = "circle";
+
+    /**
+     * Contains the tag name for group elements, {@value}.
+     */
+    protected static final String TAG_G = "g";
```
[the remark this answers is not in the document]
Comment (jakobvogel) [reply src/main/java/sirius/biz/charts/BaseChart.java]: Renamed in 5e2cbec6bd324f7b3b9f9894a056e5ffd5df4949

File: src/main/java/sirius/biz/charts/Dataset.java [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    public SequencedSet<String> getLabels() {
+        return slices.sequencedKeySet();
+    }
+
+    public List<Slice> getSlices() {
+        return List.copyOf(slices.values());
+    }
+
+    /**
+     * Represents a slice of the datset.
```
[the remark this answers is not in the document]
Comment (jakobvogel) [reply src/main/java/sirius/biz/charts/Dataset.java:207]: This class was originally part of the pie chart, where I called it like that. When refactoring it into this class, i somewhat liked the name, as I consider myself a data butcher chopping information into slices. 😜

File: src/main/java/sirius/biz/charts/PieChart.java (hunk 1 above)
Comment (jakobvogel) [reply src/main/java/sirius/biz/charts/PieChart.java]: That's a matter of taste 🙃 … But here you go: 5d944a5f2fe2c50ef6bfcb0a59cbb7e00af8a206

File: src/main/java/sirius/biz/charts/Charts.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import sirius.web.templates.pdf.TagliatellePDFContentHandler;
+
+import java.awt.Dimension;
+import java.io.StringWriter;
+
+/**
+ * Provides helpers to work with charts.
+ */
+@Register(classes = Charts.class)
+public class Charts {
```
[the remark this answers is not in the document]
Comment (sabieber) [reply src/main/java/sirius/biz/charts/Charts.java:24]: My idea was that they would not be static there and accessed via the `@Part Charts charts`. Sorry for not stating that in my initial suggestion

File: src/main/java/sirius/biz/charts/Dataset.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sabieber) [reply src/main/java/sirius/biz/charts/Dataset.java:207]: I wanted to indirectly suggest that the meaning of a slice may need a bit of explaining in the JavaDoc of the class 😉

**Review by sabieber — APPROVED** [review]
Thanks for all the actions on my nitpicking 🙏🏻💚

File: src/main/java/sirius/biz/charts/Charts.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jakobvogel) [reply src/main/java/sirius/biz/charts/Charts.java:24]: Hmm, the single surviving static method is only an internal helper anyway for creating the SVG root element with appropriately configured view box. I moved it back to the `Chart` class but removed the `static` modifier for clarification. Done in 1cd2fc0bb16101d024113f8683b142e41bf0dbe9

File: src/main/java/sirius/biz/charts/Dataset.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jakobvogel) [reply src/main/java/sirius/biz/charts/Dataset.java]: Reworked wording and documentation in 09fccfd67ab2013877d55c53cf2a37a3dfec0a3a

[the remark this answers is not in the document]
Comment (sabieber) [reply src/main/java/sirius/biz/charts/Dataset.java]: Muchas gracias

**Review by sabieber — APPROVED** [review]

**Review by jmuscireum — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Pull request merged**